### ISSUE TITLE
don't paste from excel to handsontable on mac safari browser

### ISSUE THREAD

**Issue by usernamehogehoge** [handsontable/handsontable#5121]
### Description

don't paste from excel to handsontable on mac safari browser

### Steps to reproduce

1.copy on excel
2.paste(command + v) handsontable on safari mac

### Demo
http://jsfiddle.net/hU6Kz/5703/

### Your environment
* Handsontable version:3.0.0
* Browser Name and version: safari 11.1
* Operating System:mac 10.13.4
![aa mov](https://user-images.githubusercontent.com/31416292/40547023-bea83242-606b-11e8-9979-cc0a5abbcd60.gif)

**Comment by AMBudnik** [comment]
Hi @usernamehogehoge 

I've tested it on `10.13.3` (Safari `11.03`) but pasting works as expected. Let me test it on other device.

**Comment by AMBudnik** [comment]
Can you check if [`beforePaste`](https://docs.handsontable.com/pro/3.0.0/Hooks.html#event:beforePaste) and [`afterPaste`](https://docs.handsontable.com/pro/3.0.0/Hooks.html#event:afterPaste) hooks are fired?

**Comment by usernamehogehoge** [comment]
may be No.
i am try this site.
https://docs.handsontable.com/0.16.1/tutorial-callbacks.html
![b mov](https://user-images.githubusercontent.com/31416292/40617684-0a006f24-62ca-11e8-859f-1739abcf7c8c.gif)

**Comment by AMBudnik** [comment]
If you're testing the *Callbacks* demo I recommend checking only the two hooks as when you select all there are too many of calls that it is nearly impossible to track them all and spot what we're interested on.

**Comment by usernamehogehoge** [comment]
i do not find 「beforePaste」 and 「afterPaste」 in this page 「Choose events to be logged:」
https://docs.handsontable.com/0.16.1/tutorial-callbacks.html

Please tell me what to select 「Choose events to be logged:」?

**Comment by AMBudnik** [comment]
You are using the old version of Handsontable (`0.16.1`) that did not have this two hooks.
Here's a new address https://docs.handsontable.com/3.0.0/tutorial-using-callbacks.html

**Comment by usernamehogehoge** [comment]
![aa mov](https://user-images.githubusercontent.com/31416292/40725199-c5f0cffa-645d-11e8-9725-d6a3d5b26d27.gif)

safari
1311 @203.467 [beforePaste] [[""]], [object Array],
1312 @203.527 [afterPaste] [[""]], [object Array],

chrome
18 @94.979 [afterSelection] 0, 2, 0, 2, [object Object], 0
19 @95.861 [afterChange] [object Array], "CopyPaste.paste",
20 @95.862 [afterSelection] 0, 2, 2, 2, [object Object], 0

**Comment by dbucardi** [comment]
Hi, I've tested it on 10.13.4 (Safari 11.1), pasting is not working. The paste object event.clipboardData does not have the 'text/plain' content type when the copy is done using Excel or Number, it only has "Files" e "text/html" content type, I think this is why the feature stopped working.

**Comment by AMBudnik** [comment]
Thank you for investigating this issue.

**Comment by jonathonjones** [comment]
I am also having this issue, using Safari Version 11.1.1 (13605.2.8) - are there any workarounds or suggestions for this?

In the release notes for 11.1 it looks like there was an update to the Clipboard API - I wonder if that ended up changing the clipboard data to text/html?

[Safari Release Notes](https://developer.apple.com/library/archive/releasenotes/General/WhatsNewInSafari/Articles/Safari_11_1.html#//apple_ref/doc/uid/TP40014305-CH14-SW3)

**Comment by dbucardi** [comment]
A workaround should be:

- use the hook beforePaste to get the text/html from window.event.clipboardData 
- use jquery to parse the data from html to a JavaScript array
- update the parameter data from beforePaste event with the new values parsed from html

There is some examples about using beforePaste hook on handsontable docs.

**Comment by phoenix-down** [comment]
I'm having the same issue pasting from Excel or Numbers to Safari 11.1.1.

Here's when I'm trying to paste a 3x2 grid of numbers from Excel for Mac version 16.14.1:

![screen shot 2018-07-09 at 11 44 39 pm](https://user-images.githubusercontent.com/20919658/42489687-11b56188-83d2-11e8-9591-031618d87d56.png)

Using this page:
https://docs.handsontable.com/3.0.0/tutorial-using-callbacks.html

![screen shot 2018-07-09 at 11 43 25 pm](https://user-images.githubusercontent.com/20919658/42489667-ecd76ca8-83d1-11e8-8c39-c406d5a0892e.png)

One workaround I found is instead of doing a standard paste, if you do **Edit -> Paste and Match Style**, it works. Weirdly enough, it only works if done through the menu. If you use the keyboard shortcut for Paste and Match Style. It does not work.

@AMBudnik - Any ETA on a fix?

**Comment by AMBudnik** [comment]
I am sorry but there is no milestone attached to this issue yet

**Comment by phoenix-down** [comment]
I've created a workaround in case anyone else stumbles into this thread. I've verified that it works with pasting from Excel and Numbers in Safari 11.1.1 and Handsontable 3.0.0.

```

beforePaste: (data, coords) => {
        // NOTE: This is a workaround for Safari (11.1) for pasting data from Excel and Numbers.

        // Get the paste data from the system clipboard, which is in HTML. We then need to parse the HTML
        // and get the actual values without the HTML tags.
        let parser = new DOMParser();
        let html = parser.parseFromString(window.event.clipboardData.getData('text/html'), 'text/html');
        let htmlTableRow = html.getElementsByTagName('tr');

        // Check to make sure that the clipboard contains HTML data and that Handsontable's paste data
        // structure only contains a single element of zero-length string. Since pasting works properly
        // in Chrome, the following code will only run in Safari.
        if (data.length === 1 && data[0][0] === '' && htmlTableRow.length > 0) {
          // Remove the original data structure
          data.splice(0, 1);

          // Iterate through the HTML tree and get the actual values
          for (let i = 0; i < htmlTableRow.length; i++) {
            let pasteCols = htmlTableRow[i].getElementsByTagName('td');
            let curRowData = [];
            for (let j = 0; j < pasteCols.length; j++) {
              // We use innerText here instead of innerHTML because at this level of the HTML tree, Excel
              // data contains the plaintext value, while Numbers wraps the value further in an addiitonal
              // "<p><font>...</font></p>". We really only care about the plaintext value inside.
              curRowData.push(pasteCols[j].innerText);
            }

            // Update the paste data
            data.push(curRowData);
          }
        }
      },

```

**Comment by mill3rk** [comment]
Not sure of this helps anyone anymore, but it seems to work in releases 0.33.0 to 0.34.4 then stops working outside of those releases.

**Comment by AMBudnik** [comment]
Thank you @mill3rk for checking and reporting that. It can help us to find a commit that changed the behavior

**Comment by swistach** [comment]
This issue will be solved within #4931
